components/style/matching.rs

```diff
@@ -677,6 +680,20 @@ pub trait ElementMatchMethods : TElement {
             relations |= AFFECTED_BY_PSEUDO_ELEMENTS;
         }
 
+        // Now we might need to update the parent flags regarding which
+        // selectors are matched.
+        let (child_flags, parent_flags) = ElementFlags::from_relations(relations);
+        if !child_flags.is_empty() {
+            // FIXME(emilio): This doesn't need to be atomic.
```

Comment:
It seems like we either need sequential consistency or we need to use the atomic insert that we use for the children. Given that the latter is probably cheaper, it's not clear to me what we could optimize here.